### PR TITLE
Replace dynamically-loaded kernel32 functions with winapi

### PR DESCRIPTION
**Pull request by sterlingjensen**
The libloaderapi implementation pre-dates availability in consoleapi (0.3.7).
The winapi feature flags have fallen out of sync with what is in use, as no relative paths to the crate are used in this project, so they were removed.

### CODE REVIEW

**Comment by chrisduerr** [comment]
> The winapi feature flags have fallen out of sync with what is in use, as no relative paths to the crate are used in this project, so they were removed.

I'm not sure what you mean by this? What do feature flags have to do with relative paths?

If they're not necessary that's fine, though it might be worthwhile to check if we might want to do no-default-features and then add only what we need ourselves.

**Comment by davidhewitt** [comment]
As much as I love this, I'm not sure we can do it. Won't it break support for Windows versions which predate the existence of the ConPTY api?

For example, on all Windows 7 versions `kernel32` doesn't have these functions. So with this PR I expect Alacritty would crash during load-time dynamic linking.

(Ofc I could just be totally misunderstanding how things work. Correct me if so.)

Alternatively; go ahead and do it. It would make having ConPTY support a hard requirement for Alacritty on Windows. We could then take out the Winpty implementation completely and simplify the lot 😈 🙈

**Comment by sterlingjensen** [comment]
I've only seen the feature flags used to control what is made available in a `mod.rs` import (`winapi::um`), with sub-modules being controlled in the `Cargo.toml` (...`features = ["d3d11"]`...). This means that all usage would rely on paths relative to that import (`um::d3d11::D3D11_RECT`). I don't see any default features in `winapi`.
I hadn't considered graceful failures for Windows 7... that would certainly make this a non-starter. I'll spin up a vm and check it out.

**Comment by davidhewitt** [comment]
> I hadn't considered graceful failures for Windows 7... that would certainly make this a non-starter. I'll spin up a vm and check it out.

To be fair, Windows 7's end of life is next month, but there's still Windows 8 and old versions of Windows 10 kicking around :)

**Comment by chrisduerr** [comment]
@davidhewitt Thanks for pointing that out, very, very good point. I don't think we should be killing off WinPTY just yet.

**Comment by davidhewitt** [comment]
A quick thought on this - if we decide not to kill off the dynamic loading right now, maybe let's update the comment so that people don't try this again until we're ready for it?

**Comment by sterlingjensen** [comment]
Well that was a frustrating experience, the approaching EOL really shows when it comes to tooling. Long story short: now I know how this seemingly low hanging fruit got overlooked until I dove for it: this change breaks things. I can think of a bunch of ways to fix this, but the only correct way I can imagine involves some high level design choices inside `tty::windows`. I think that discussion is better had at #3085.
Should I close this, change the title to include '[WIP]', or leave it as is for now?

**Comment by sterlingjensen** [comment]
@davidhewitt That would have saved me a day, so it gets my vote :smirk:

**Comment by chrisduerr** [comment]
>  Should I close this, change the title to include '[WIP]', or leave it as is for now?

I'm in favor of closing this, yeah. Having a ton of PRs open which aren't going to see any work done makes things a lot harder for me.

**Comment by davidhewitt** [comment]
> @davidhewitt That would have saved me a day, so it gets my vote 😏

Ah damn, sorry about that (the original dynamic load was written by me, so I feel responsible for that) ... :/

**Comment by davidhewitt** [comment]
>  I can think of a bunch of ways to fix this, but the only correct way I can imagine involves some high level design choices inside tty::windows. I think that discussion is better had at #3085.

Please do drop your thoughts there. Very keen to hear any ideas / feedback for tidying this up :)

**Comment by sterlingjensen** [comment]
I just pushed a patch that leaves the removal of the feature flags, but reverts the changes to dynamic loading and adds a todo comment to ward off others. @chrisduerr let me know if that works as a nice way to close through a merge, otherwise I'll close it.

@davidhewitt Don't feel bad, I almost messaged you when I first noticed the opportunity to drop the dynamic load, and saw your involvement in both it and the winapi implementation that would replace it :) But I resisted the urge, because it felt like volunteering you for more work. Anyway, it wasn't a wasted day: I've now got an up to date golden master for win7, some leads on some potential gl-rs bugs, and a new appreciation for not using Windows 7 as a daily driver.

**Comment by davidhewitt** [comment]
> I just pushed a patch that leaves the removal of the feature flags

Just curious about the winapi feature flags, how does it still build without them? I guess we're probably getting lucky that these features are already pulled in by upstream crates?

**Comment by sterlingjensen** [comment]
It doesn't take much, with all the interdependency - winapi directly pulls in most of it, and its rest likely from its other dependencies. I've resynced the feature flags.

**Comment by davidhewitt** [comment]
:+1: thanks, that seems like a good idea to me

File: alacritty_terminal/Cargo.toml [hunk 1]
```diff
@@ -35,9 +35,12 @@ winpty = { path = "../winpty" }
 mio-named-pipes = "0.1"
 miow = "0.3"
 dunce = "1.0"
-winapi = { version = "0.3.7", features = ["impl-default", "winuser", "synchapi", "roerrorapi", "winerror", "wincon", "wincontypes"]}
 widestring = "0.4"
 mio-anonymous-pipes = "0.1"
+[target.'cfg(windows)'.dependencies.winapi]
+version = "0.3.7"
+features = ["impl-default", "basetsd", "libloaderapi", "minwindef", "ntdef", "processthreadsapi",
+            "winbase", "wincon", "wincontypes", "winerror", "winnt", "winuser"]
```
Comment (chrisduerr) [inline alacritty_terminal/Cargo.toml]: Why did you add new feature flags that previously weren't required at all?

I also think that if we need to break it over multiple lines anyways, we might as well stick with the old version of defining `winapi` as a single element instead of a nested field.

**Comment by davidhewitt** [comment]
> Why did you add new feature flags that previously weren't required at all?

I guess that these are all the `winapi` submodules that appear in paths inside `alacritty_terminal`. Previously some of these features must have been pulled in by dependencies (or automatically as dependencies of other `winapi` features).

I'm not sure what the correct pattern is here either. It feels most correct to list them all out here as @sterlingjensen has done. But for the sake of brevity, and to minimise maintenance burden keeping the list in sync, I tend to go for the smallest list of features which are needed to get a successful compile. As we saw here that's actually  no feature flags :smile:

**Comment by chrisduerr** [comment]
I'm all in favor of adding all depedencies we do have ourselves. I just want to know.

File: alacritty_terminal/Cargo.toml (hunk 1 above)
Comment (sterlingjensen) [reply alacritty_terminal/Cargo.toml]: They were in use before, but not specified in the manifest - out of sync.
One single line toml table, here comes the line wrap :metal:

**Comment by chrisduerr** [comment]
> unless they are valid within a value

Well, they are valid within arrays, so there will be no problem. As you can see, there are no newlines outside of the array. That's why the closing `]` and `}` are on the same line.

**Comment by sterlingjensen** [comment]
Oh I'm certainly in favor of doing the thing that doesn't cause a line wrap, I'm just glad that you did it - and not me :grinning:
